gustaf/spline/microstructure/generator.py

```diff
@@ -0,0 +1,553 @@
+                        for j in range(1, self.tiling[i_pd])
+                ]
+                # insert knots in both the deformation function
+                deformation_function_copy_.insert_knots(i_pd, new_knots)
+        else:
+            self._logd(
+                    "New knots will be inserted one by one with the objective"
+                    " to evenly distribute tiles within the parametric domain"
+            )
+            for i_pd in range(deformation_function_copy_.para_dim):
+                n_current_spans = (len(ukvs[i_pd]) - 1)
+                if self.tiling[i_pd] == n_current_spans:
+                    continue
+                elif self.tiling[i_pd] < n_current_spans:
+                    self._logw(
+                            f"The requested tiling can not be provided, as "
+                            f"there are too many knotspans in the deformation"
+                            f" function. The tiling in parametric dimension "
+                            f"{i_pd} will be set to {n_current_spans}"
+                    )
+                    self.tiling[i_pd] = n_current_spans
+                else:
+                    # Determine new knots
+                    n_k_span = np.zeros(n_current_spans, dtype=int)
+                    span_measure = np.diff(ukvs[i_pd])
+                    for _ in range(self.tiling[i_pd] - n_current_spans):
+                        add_knot = np.argmax(span_measure)
+                        n_k_span[add_knot] += 1
+                        span_measure[add_knot] *= n_k_span[add_knot] / (
+                                n_k_span[add_knot] + 1
+                        )
+
+                    new_knots = []
+                    for i, nks in enumerate(n_k_span):
+                        new_knots.extend(
+                                np.linspace(
+                                        ukvs[i_pd][i], ukvs[i_pd][i + 1],
+                                        nks + 2
+                                )[1:-1]
+                        )
+                    deformation_function_copy_.insert_knots(i_pd, new_knots)
+
+        # Bezier Extraction for composition
+        def_fun_patches = deformation_function_copy_.extract.beziers()
+
+        # Calculate parametric space representation for parametrized
+        # microstructures
+        is_parametrized = self.parametrization_function is not None
+        if is_parametrized:
+            para_space_dimensions = [[u[0], u[-1]] for u in ukvs]
+            # Trust me @j042
+            def_fun_para_space = base.Bezier(
+                    degrees=[1] * deformation_function_copy_.para_dim,
+                    control_points=np.array(
+                            list(
+                                    itertools.product(
+                                            *para_space_dimensions[::-1]
+                                    )
+                            )
+                    )[:, ::-1]
+            ).bspline
```

Comment:
is this to save an effort to write kvs?

Reply:
no that basically creates a box-bspline that describes the parametric space of the deformation function